TestBinaryShiftOperation test fails

```
got ./...
panic triggered: test
interrupt
interrupt
~~~ FAIL: (Terst)
	runtime_test.go:592:
		FAIL (==)
		     got: 1073741823 (int32)
		expected: -1073741824 (int)
--- FAIL: TestBinaryShiftOperation (0.00s)
```

Fails on 64bit OSX:

```
uname -a
Darwin SkyNetM1.local 22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:30 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8103 arm64


go test -run=TestBinaryShiftOperation -v
Alias tip: got -run=TestBinaryShiftOperation -v
=== RUN   TestBinaryShiftOperation
~~~ FAIL: (Terst)
	runtime_test.go:592:
		FAIL (==)
		     got: 1073741823 (int32)
		expected: -1073741824 (int)
--- FAIL: TestBinaryShiftOperation (0.00s)
FAIL
exit status 1
FAIL	github.com/robertkrimen/otto	0.139s
```

```
=== RUN   Test_issue457
    issue2_test.go:20:
        	Error Trace:	/Users/andig/htdocs/otto/issue2_test.go:20
        	Error:      	Not equal:
        	            	expected: "2147483647,-2147483648,-1073741824"
        	            	actual  : "2147483647,-2147483648,1073741823"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-2147483647,-2147483648,-1073741824
        	            	+2147483647,-2147483648,1073741823
        	Test:       	Test_issue457
--- FAIL: Test_issue457 (0.00s)
```

```
=== RUN   Test_issue457
shift1: otto.Value{kind:2, value:-2.147483648e+09}, otto.Value{kind:2, value:1}
shift2: 2147483647, 1
shift3: 1073741823
    issue2_test.go:20:
        	Error Trace:	issue2_test.go:20
        	Error:      	Not equal:
        	            	expected: "2147483647,-2147483648,-1073741824"
        	            	actual  : "2147483647,-2147483648,1073741823"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-2147483647,-2147483648,-1073741824
        	            	+2147483647,-2147483648,1073741823
        	Test:       	Test_issue457
--- FAIL: Test_issue457 (0.00s)
```

```
=== RUN   Test_issue457
shift1: otto.Value{kind:2, value:-2.147483648e+09}, otto.Value{kind:2, value:1}
shift2: 2147483647, 1
shift3: 1073741823
    issue2_test.go:22:
        	Error Trace:	issue2_test.go:22
        	Error:      	Not equal:
        	            	expected: "2147483647,-2147483648,-1073741824"
        	            	actual  : "2147483647,-2147483648,1073741823"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-2147483647,-2147483648,-1073741824
        	            	+2147483647,-2147483648,1073741823
        	Test:       	Test_issue457
--- FAIL: Test_issue457 (0.00s)
=== RUN   Test_issue457_toInt32
val: -2147483648, fval: -2.147483648e+09
rem1: -2.147483648e+09, float_2_32: 4.294967296e+09
rem3: 2.147483648e+09
rem4: 2.147483648e+09, float_2_31: 2.147483648e+09
    issue2_test.go:47:
        	Error Trace:	issue2_test.go:47
        	Error:      	Not equal:
        	            	expected: -2147483648
        	            	actual  : 2147483647
        	Test:       	Test_issue457_toInt32
--- FAIL: Test_issue457_toInt32 (0.00s)
```

What I don't get: why aren't you seeing this error on CI? The runners are 64bits afaik. What is different between those and my local machine?

> My machine here is 64bit but x86_64, I thought this was a 32bit arch issue but it could be something more subtle to do with internal architecture.

Interesting. Had a quick look at the code and didn't see anything obvious. Still need to look at the issue.

Now:
```
=== RUN   Test_issue457_toInt32_2
maxInt32: 2147483647 minInt32: -2147483648
float_2_32: 4294967296, float_2_31: 2147483648
val: -2147483648, fval: -2.147483648e+09
rem1: -2.147483648e+09, float_2_32: 4.294967296e+09
rem3: 2.147483648e+09
rem4: 2.147483648e+09, float_2_31: 2.147483648e+09
out2: 2.147483648e+09, float64, 2147483647
    issue2_test.go:75:
        	Error Trace:	issue2_test.go:75
        	Error:      	Not equal:
        	            	expected: -2147483648
        	            	actual  : 2147483647
        	Test:       	Test_issue457_toInt32_2
        	Messages:   	out2
--- FAIL: Test_issue457_toInt32_2 (0.00s)
```

And:

<img width="326" alt="Screenshot 2022-11-30 at 13 07 14" src="https://user-images.githubusercontent.com/184815/204792163-c56d0b1b-305c-4d3a-ba30-d4e655cafdda.png">

```
=== RUN   Test_issue457_toInt32_3
abs: 2.147483648e+09
iv: 2.147483648e+09
iv32: 2.147483648e+09
res1: -2147483648
--- FAIL: Test_issue457_toInt32_3 (0.00s)
```

I guess you've fixed the original one? Here's my entire output:

```
=== RUN   Test_issue457
    issue2_test.go:23: 2147483647,-2147483648,-1073741824
--- PASS: Test_issue457 (0.00s)
=== RUN   Test_issue457_toInt32
val: -2147483648, fval: -2.147483648e+09
rem1: -2.147483648e+09, float_2_32: 4.294967296e+09
rem3: 2.147483648e+09
rem4: 2.147483648e+09, float_2_31: 2.147483648e+09
    issue2_test.go:47:
        	Error Trace:	/Users/andig/htdocs/otto/issue2_test.go:47
        	Error:      	Not equal:
        	            	expected: -2147483648
        	            	actual  : 2147483647
        	Test:       	Test_issue457_toInt32
--- FAIL: Test_issue457_toInt32 (0.00s)
=== RUN   Test_issue457_toInt32_2
maxInt32: 2147483647 minInt32: -2147483648
float_2_32: 4294967296, float_2_31: 2147483648
val: -2147483648, fval: -2.147483648e+09
rem1: -2.147483648e+09, float_2_32: 4.294967296e+09
rem3: 2.147483648e+09
rem4: 2.147483648e+09, float_2_31: 2.147483648e+09
out2: 2.147483648e+09, float64, 2147483647
    issue2_test.go:75:
        	Error Trace:	/Users/andig/htdocs/otto/issue2_test.go:75
        	Error:      	Not equal:
        	            	expected: -2147483648
        	            	actual  : 2147483647
        	Test:       	Test_issue457_toInt32_2
        	Messages:   	out2
--- FAIL: Test_issue457_toInt32_2 (0.00s)
=== RUN   Test_issue457_toInt32_3
abs: 2.147483648e+09
iv: 2.147483648e+09
iv32: 2.147483648e+09
res1: -2147483648
--- FAIL: Test_issue457_toInt32_3 (0.00s)
```

You're not gonna like it:

```
got ./...
~~~ FAIL: (Terst)
	issue_test.go:639:
		FAIL (==)
		     got: 000000000626540e598e480000000000000000004048391f4a1929274d7c695c (*otto._object=ptr)
		expected: aa747c502a898200f9e4fa21bac68136f886a0e27aec70ba06daf2e2a5cb5597
--- FAIL: Test_issue87 (0.01s)
panic triggered: test
interrupt
interrupt
~~~ FAIL: (Terst)
	runtime_test.go:593:
		FAIL (==)
		     got: 0 (uint32)
		expected: 1073741824 (int)
--- FAIL: TestBinaryShiftOperation (0.00s)
~~~ FAIL: (Terst)
	string_test.go:67:
		FAIL (==)
		     got: 0 (uint16)
		expected: 65535 (int)
--- FAIL: TestString_fromCharCode (0.00s)
~~~ FAIL: (Terst)
	value_test.go:174:
		FAIL (==)
		     got: 0 (uint32)
		expected: 2147483647 (uint32)
	value_test.go:174:
		FAIL (==)
		     got: 0 (uint32)
		expected: 4294967295 (uint32)
	value_test.go:174:
		FAIL (==)
		     got: 0 (uint32)
		expected: 1 (uint32)
--- FAIL: Test_toUint32 (0.00s)
~~~ FAIL: (Terst)
	value_test.go:196:
		FAIL (==)
		     got: 0 (uint16)
		expected: 65535 (uint16)
	value_test.go:196:
		FAIL (==)
		     got: 0 (uint16)
		expected: 65535 (uint16)
	value_test.go:196:
		FAIL (==)
		     got: 0 (uint16)
		expected: 1 (uint16)
--- FAIL: Test_toUint16 (0.00s)
FAIL
FAIL	github.com/robertkrimen/otto	1.214s
ok  	github.com/robertkrimen/otto/ast	0.256s
?   	github.com/robertkrimen/otto/dbg	[no test files]
?   	github.com/robertkrimen/otto/file	[no test files]
?   	github.com/robertkrimen/otto/otto	[no test files]
ok  	github.com/robertkrimen/otto/parser	0.175s
?   	github.com/robertkrimen/otto/registry	[no test files]
?   	github.com/robertkrimen/otto/repl	[no test files]
?   	github.com/robertkrimen/otto/terst	[no test files]
?   	github.com/robertkrimen/otto/test	[no test files]
?   	github.com/robertkrimen/otto/token	[no test files]
?   	github.com/robertkrimen/otto/underscore	[no test files]
FAIL
```

at

```
commit 3376c332e8f9c4e21cc2b41b29c85635ff7b7d7b (HEAD -> fix/number-to-ints, origin/fix/number-to-ints)
Author: Steven Hartland <steven.hartland@multiplay.co.uk>
Date:   Wed Nov 30 15:17:41 2022 +0000

    fix: leverage go casts for integer conversion

    Leverage the power of go casts to simplify the conversion from float64
    to integer types.

 value_number.go | 19 +++----------------
 1 file changed, 3 insertions(+), 16 deletions(-)
```

No. Same result, but we also didn't test these before:

```
got ./...
~~~ FAIL: (Terst)
	issue_test.go:639:
		FAIL (==)
		     got: 000000000626540e598e480000000000000000004048391f4a1929274d7c695c (*otto._object=ptr)
		expected: aa747c502a898200f9e4fa21bac68136f886a0e27aec70ba06daf2e2a5cb5597
--- FAIL: Test_issue87 (0.01s)
panic triggered: test
interrupt
interrupt
~~~ FAIL: (Terst)
	runtime_test.go:593:
		FAIL (==)
		     got: 0 (uint32)
		expected: 1073741824 (int)
--- FAIL: TestBinaryShiftOperation (0.00s)
~~~ FAIL: (Terst)
	string_test.go:67:
		FAIL (==)
		     got: 0 (uint16)
		expected: 65535 (int)
--- FAIL: TestString_fromCharCode (0.00s)
~~~ FAIL: (Terst)
	value_test.go:174:
		FAIL (==)
		     got: 0 (uint32)
		expected: 2147483647 (uint32)
	value_test.go:174:
		FAIL (==)
		     got: 0 (uint32)
		expected: 4294967295 (uint32)
	value_test.go:174:
		FAIL (==)
		     got: 0 (uint32)
		expected: 1 (uint32)
--- FAIL: Test_toUint32 (0.00s)
~~~ FAIL: (Terst)
	value_test.go:196:
		FAIL (==)
		     got: 0 (uint16)
		expected: 65535 (uint16)
	value_test.go:196:
		FAIL (==)
		     got: 0 (uint16)
		expected: 65535 (uint16)
	value_test.go:196:
		FAIL (==)
		     got: 0 (uint16)
		expected: 1 (uint16)
--- FAIL: Test_toUint16 (0.00s)
FAIL
FAIL	github.com/robertkrimen/otto	1.101s
ok  	github.com/robertkrimen/otto/ast	(cached)
?   	github.com/robertkrimen/otto/dbg	[no test files]
?   	github.com/robertkrimen/otto/file	[no test files]
?   	github.com/robertkrimen/otto/otto	[no test files]
ok  	github.com/robertkrimen/otto/parser	(cached)
?   	github.com/robertkrimen/otto/registry	[no test files]
?   	github.com/robertkrimen/otto/repl	[no test files]
?   	github.com/robertkrimen/otto/terst	[no test files]
?   	github.com/robertkrimen/otto/test	[no test files]
?   	github.com/robertkrimen/otto/token	[no test files]
?   	github.com/robertkrimen/otto/underscore	[no test files]
FAIL
```

at

```
commit f062ffba5edb28c8e1526ef15a6f1207ea51c83c (HEAD -> fix/number-to-ints)
Author: Steven Hartland <steven.hartland@multiplay.co.uk>
Date:   Wed Nov 30 15:11:54 2022 +0000

    fix: sign and overflow

    Fix sign and overflow / underflow cast issues.
```

--- a/value_number.go
+++ b/value_number.go
@@ -251,8 +251,8 @@ func toUint32(value Value) uint32 {
 		return 0
 	}
 
-	// Convert to uint64 before uint32 to force correct wrapping.
-	return uint32(uint64(floatValue))
+	// Convert to int64 before uint32 to force correct wrapping.
+	return uint32(int64(floatValue))
 }
 
 // ECMA 262 - 6.0 - 7.1.8.
@@ -271,8 +271,8 @@ func toUint16(value Value) uint16 {
 		return 0
 	}
 
-	// Convert to uint64 before uint16 to force correct wrapping.
-	return uint16(uint64(floatValue))
+	// Convert to int64 before uint16 to force correct wrapping.
+	return uint16(int64(floatValue))
 }
 
 // toIntSign returns sign of a number converted to -1, 0 ,1

--- a/value_number.go
+++ b/value_number.go
@@ -228,13 +228,15 @@ func toInt32(value Value) int32 {
 		return 0
 	}
 
-	iv := math.Floor(math.Abs(floatValue))
+	iv := math.Copysign(math.Floor(math.Abs(floatValue)), floatValue)
 	iv32 := math.Mod(iv, float_2_32)
+
 	if iv32 >= float_2_31 {
-		return int32(iv32 - float_2_32)
+		iv32 = iv32 - float_2_32
 	}
 
-	return int32(iv32)
+	// Convert to int64 before int32 to force correct wrapping.
+	return int32(int64(iv32))
 }
 
 func toUint32(value Value) uint32 {
@@ -256,10 +258,11 @@ func toUint32(value Value) uint32 {
 		return 0
 	}
 
-	iv := math.Floor(math.Abs(floatValue))
+	iv := math.Copysign(math.Floor(math.Abs(floatValue)), floatValue)
 	iv32 := math.Mod(iv, float_2_32)
 
-	return uint32(iv32)
+	// Convert to uint64 before uint32 to force correct wrapping.
+	return uint32(uint64(iv32))
 }
 
 // ECMA 262 - 6.0 - 7.1.8.
@@ -278,10 +281,11 @@ func toUint16(value Value) uint16 {
 		return 0
 	}
 
-	iv := math.Floor(math.Abs(floatValue))
+	iv := math.Copysign(math.Floor(math.Abs(floatValue)), floatValue)
 	iv16 := math.Mod(iv, float_2_16)
 
-	return uint16(iv16)
+	// Convert to uint64 before uint16 to force correct wrapping.
+	return uint16(uint64(iv16))
 }
 
 // toIntSign returns sign of a number converted to -1, 0 ,1

--- a/value_number.go
+++ b/value_number.go
@@ -228,15 +228,8 @@ func toInt32(value Value) int32 {
 		return 0
 	}
 
-	iv := math.Copysign(math.Floor(math.Abs(floatValue)), floatValue)
-	iv32 := math.Mod(iv, float_2_32)
-
-	if iv32 >= float_2_31 {
-		iv32 = iv32 - float_2_32
-	}
-
 	// Convert to int64 before int32 to force correct wrapping.
-	return int32(int64(iv32))
+	return int32(int64(floatValue))
 }
 
 func toUint32(value Value) uint32 {
@@ -258,11 +251,8 @@ func toUint32(value Value) uint32 {
 		return 0
 	}
 
-	iv := math.Copysign(math.Floor(math.Abs(floatValue)), floatValue)
-	iv32 := math.Mod(iv, float_2_32)
-
 	// Convert to uint64 before uint32 to force correct wrapping.
-	return uint32(uint64(iv32))
+	return uint32(uint64(floatValue))
 }
 
 // ECMA 262 - 6.0 - 7.1.8.
@@ -281,11 +271,8 @@ func toUint16(value Value) uint16 {
 		return 0
 	}
 
-	iv := math.Copysign(math.Floor(math.Abs(floatValue)), floatValue)
-	iv16 := math.Mod(iv, float_2_16)
-
 	// Convert to uint64 before uint16 to force correct wrapping.
-	return uint16(uint64(iv16))
+	return uint16(uint64(floatValue))
 }
 
 // toIntSign returns sign of a number converted to -1, 0 ,1

--- a/value_number.go
+++ b/value_number.go
@@ -262,6 +262,28 @@ func toUint32(value Value) uint32 {
 	return uint32(iv32)
 }
 
+// ECMA 262 - 6.0 - 7.1.8.
+func toUint16(value Value) uint16 {
+	switch value := value.value.(type) {
+	case int8:
+		return uint16(value)
+	case uint8:
+		return uint16(value)
+	case uint16:
+		return value
+	}
+
+	floatValue := value.float64()
+	if math.IsNaN(floatValue) || math.IsInf(floatValue, 0) || floatValue == 0 {
+		return 0
+	}
+
+	iv := math.Floor(math.Abs(floatValue))
+	iv16 := math.Mod(iv, float_2_16)
+
+	return uint16(iv16)
+}
+
 // toIntSign returns sign of a number converted to -1, 0 ,1
 func toIntSign(value Value) int {
 	switch value := value.value.(type) {

--- a/value_number.go
+++ b/value_number.go
@@ -214,91 +214,52 @@ func (value Value) number() (number _number) {
 
 // ECMA 262: 9.5
 func toInt32(value Value) int32 {
-	{
-		switch value := value.value.(type) {
-		case int8:
-			return int32(value)
-		case int16:
-			return int32(value)
-		case int32:
-			return value
-		}
+	switch value := value.value.(type) {
+	case int8:
+		return int32(value)
+	case int16:
+		return int32(value)
+	case int32:
+		return value
 	}
+
 	floatValue := value.float64()
-	if math.IsNaN(floatValue) || math.IsInf(floatValue, 0) {
+	if math.IsNaN(floatValue) || math.IsInf(floatValue, 0) || floatValue == 0 {
 		return 0
 	}
-	if floatValue == 0 { // This will work for +0 & -0
-		return 0
+
+	iv := math.Floor(math.Abs(floatValue))
+	iv32 := math.Mod(iv, float_2_32)
+	if iv32 >= float_2_31 {
+		return int32(iv32 - float_2_32)
 	}
-	remainder := math.Mod(floatValue, float_2_32)
-	if remainder > 0 {
-		remainder = math.Floor(remainder)
-	} else {
-		remainder = math.Ceil(remainder) + float_2_32
-	}
-	if remainder > float_2_31 {
-		return int32(remainder - float_2_32)
-	}
-	return int32(remainder)
+
+	return int32(iv32)
 }
 
 func toUint32(value Value) uint32 {
-	{
-		switch value := value.value.(type) {
-		case int8:
-			return uint32(value)
-		case int16:
-			return uint32(value)
-		case uint8:
-			return uint32(value)
-		case uint16:
-			return uint32(value)
-		case uint32:
-			return value
-		}
+	switch value := value.value.(type) {
+	case int8:
+		return uint32(value)
+	case int16:
+		return uint32(value)
+	case uint8:
+		return uint32(value)
+	case uint16:
+		return uint32(value)
+	case uint32:
+		return value
 	}
-	floatValue := value.float64()
-	if math.IsNaN(floatValue) || math.IsInf(floatValue, 0) {
-		return 0
-	}
-	if floatValue == 0 {
-		return 0
-	}
-	remainder := math.Mod(floatValue, float_2_32)
-	if remainder > 0 {
-		remainder = math.Floor(remainder)
-	} else {
-		remainder = math.Ceil(remainder) + float_2_32
-	}
-	return uint32(remainder)
-}
 
-func toUint16(value Value) uint16 {
-	{
-		switch value := value.value.(type) {
-		case int8:
-			return uint16(value)
-		case uint8:
-			return uint16(value)
-		case uint16:
-			return value
-		}
-	}
 	floatValue := value.float64()
-	if math.IsNaN(floatValue) || math.IsInf(floatValue, 0) {
+	if math.IsNaN(floatValue) || math.IsInf(floatValue, 0) || floatValue == 0 {
 		return 0
 	}
-	if floatValue == 0 {
-		return 0
-	}
-	remainder := math.Mod(floatValue, float_2_16)
-	if remainder > 0 {
-		remainder = math.Floor(remainder)
-	} else {
-		remainder = math.Ceil(remainder) + float_2_16
-	}
-	return uint16(remainder)
+
+	iv := math.Floor(math.Abs(floatValue))
+	iv32 := math.Mod(iv, float_2_32)
+
+	return uint32(iv32)
 }
 
 // toIntSign returns sign of a number converted to -1, 0 ,1

